Zend/zend_vm_def.h

```diff
@@ -451,25 +451,102 @@ ZEND_VM_COLD_CONSTCONST_HANDLER(16, ZEND_IS_IDENTICAL, CONST|TMP|VAR|CV, CONST|T
 	SAVE_OPLINE();
 	op1 = GET_OP1_ZVAL_PTR_DEREF(BP_VAR_R);
 	op2 = GET_OP2_ZVAL_PTR_DEREF(BP_VAR_R);
-	result = fast_is_identical_function(op1, op2);
+	if (Z_TYPE_P(op1) != Z_TYPE_P(op2)) {
+		/* They are not identical, return false. Note that this has to check for undefined variable errors when IS_NULL is possible. */
+		ZEND_VM_C_LABEL(is_different_nothrow):
+		FREE_OP1();
+		FREE_OP2();
+		ZEND_VM_SMART_BRANCH(0, 1);
+		return;
+	}
+	if (Z_TYPE_P(op1) <= IS_TRUE) {
+		/* They are identical, return true */
+		FREE_OP1();
+		FREE_OP2();
+		ZEND_VM_SMART_BRANCH(1, 1);
+		return;
+	}
+	switch (Z_TYPE_P(op1)) {
+		case IS_LONG:
+			result = (Z_LVAL_P(op1) == Z_LVAL_P(op2));
+			break;
+		case IS_RESOURCE:
+			result = (Z_RES_P(op1) == Z_RES_P(op2));
+			break;
+		case IS_DOUBLE:
+			result = (Z_DVAL_P(op1) == Z_DVAL_P(op2));
+			break;
+		case IS_STRING:
+			result = zend_string_equals(Z_STR_P(op1), Z_STR_P(op2));
+			break;
+		case IS_ARRAY:
+			/* This may cause EG(exception) due to infinite nesting, but other zval types won't? */
+			/* XXX hash_zval_identical_function is not static */
+			result = zend_is_identical_array(Z_ARRVAL_P(op1), Z_ARRVAL_P(op2));
+			break;
+		case IS_OBJECT:
+			result = (Z_OBJ_P(op1) == Z_OBJ_P(op2));
+			break;
+		default:
+			ZEND_VM_C_GOTO(is_different_nothrow);
+	}
 	FREE_OP1();
 	FREE_OP2();
-	ZEND_VM_SMART_BRANCH(result, 1);
+	ZEND_VM_SMART_BRANCH(result, 0);
```

Comment:
It looks like, you messed the meaning of the second argument of ZEND_VM_SMART_BRANCH().
It should be 1 here (because FREE_OP on IS_OBJECT may call destructor that throws exception) and 0 in other places.